src/domain/Filters/AddFilterDropdownMenu/AddFilterDropdownMenu.tsx

```diff
@@ -24,129 +24,179 @@ interface IAddFilterDropdownMenuProps {
+  operator?: string
+  /** List of select options to fill the select input for the filter with */
+  selectOptions?: Array<ISelectOption<FilterValue>>
+  /** A custom input component to use to select the filter. Overrides selectOptions */
+  customInputComponent?: React.ComponentType<{ onChange: (value: FilterValue | null) => void, value: FilterValue | null}>
+  /**
+   * @deprecated was never fully implemented and custom input component covers it's cases
+   */
+  type?: 'SINGLE_SELECT' | 'NUMBER'
 }
 
-interface ISelectOption {
+interface ISelectOption<FilterValue = string | number> {
   label: string
-  value: string | number
+  value: FilterValue
+}
+
+interface IAddedFilter<FilterValue> {
+  /** The filter definition that was added */
+  filter: IAddFilterDropdownMenuFilter<FilterValue>
+  /**
+   * The exact option within the filter that was added.
+   * For custom input components, the label is 'custom'
+   * as we don't want to break the interface
+   */
+  addedOption: ISelectOption<FilterValue>
 }
 
-interface IAddedFilter {
-  filter: IAddFilterDropdownMenuFilter
-  addedOption: ISelectOption
+interface IAddFilterDropdownMenuContentProps<FilterValue = string | number> {
+  filterMessage?: string
+  filters: Array<IAddFilterDropdownMenuFilter<FilterValue>>
+  onFilterAdded: (selectedFilter: IAddedFilter<FilterValue>) => void
+  closeMenu: () => void
 }
 
-const AddFilterDropdownMenu: React.FC<IAddFilterDropdownMenuProps> = ({
+function AddFilterDropdownMenu<FilterValue = string | number> ({
   filterMessage = 'Show all items where:',
   toggleComponent,
   componentContext,
   filters,
   onFilterAdded
-}) => {
-  const [filterField, setFilterField] = useState('')
-  const [filterValue, setFilterValue] = useState('')
-
-  function getFieldInputOptions (): ISelectOption[] | undefined {
-    if (filters) {
-      return filters.map((filter: IAddFilterDropdownMenuFilter) => (
-        {
-          label: filter.label,
-          value: filter.fieldName
-        })
-      )
-    }
-  }
-
-  function handleFieldChange (option: any) {
-    setFilterField(option.value)
-  }
-
-  function handleValueChange (option: any) {
-    setFilterValue(option.value)
-  }
-
-  function valueInputOption (): Array<{ label: string, value: string | number | boolean }> | undefined {
-    const selectedFilter = filters.find((filter: IAddFilterDropdownMenuFilter) => filter.fieldName === filterField)
+}: IAddFilterDropdownMenuProps<FilterValue>) {
+  return (
+    <StyledDropdownMenu
+      toggleComponent={toggleComponent}
+      componentContext={componentContext}
+    >
+      {({ closeMenu }: { closeMenu: (() => void) }) => (
+        <DropdownMenuContent<FilterValue>
+          closeMenu={closeMenu}
+          filterMessage={filterMessage}
+          filters={filters}
+          onFilterAdded={onFilterAdded}
+        />
+      )}
+    </StyledDropdownMenu>
+  )
+}
 
+function DropdownMenuContent<FilterValue = string | number> (
```

Comment:
Is splitting this for optimisation?

Reply:
makes it easier to deal with the `closeMenu` provided by the dropdown menu

Reply:
ah, so you can use it in the useCallback, makes sense.